src/main.cpp

```diff
@@ -1045,19 +1052,125 @@ bool ContextualCheckTransaction(
+                    error("ContextualCheckTransaction(): ZIP225 version too low"),
+                    REJECT_INVALID, "bad-tx-zip225-version-too-low");
+            }
+
+            if (tx.nVersion > ZIP225_MAX_TX_VERSION) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): ZIP225 version too high"),
+                    REJECT_INVALID, "bad-tx-zip225-version-too-high");
+            }
+
+            // tx.nConsensusBranchId must match the current consensus branch id
+            if (!(tx.GetConsensusBranchId() && *tx.GetConsensusBranchId() == consensusBranchId)) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): consensus branch id not set"),
+                    REJECT_INVALID, "bad-tx-missing-consensus-branch-id");
+            }
+        }
+
+        //nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
+        size_t max_spends = 1 << 16;
+        if (tx.vShieldedSpend.size() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Sapling spends"),
+                REJECT_INVALID, "bad-tx-too-many-sapling-spends");
+        }
+        if (tx.vShieldedOutput.size() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Sapling outputs"),
+                REJECT_INVALID, "bad-tx-too-many-sapling-outputs");
+        }
+        if (orchard_bundle.GetNumActions() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Orchard actions"),
+                REJECT_INVALID, "bad-tx-too-many-orchard-actions");
+        }
+
+        if (tx.IsCoinBase()) {
+            // A coinbase transaction cannot have Orchard spends.
+            if (orchard_bundle.SpendsEnabled()) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): coinbase has orchard shielded spends"),
+                    REJECT_INVALID, "bad-cb-has-orchard-spends");
+            }
+
+            // Check that Orchard coinbase outputs can be decrypted with the all-zeros OVK
+        }
+    } else {
+        // Rules that apply generally before Nu5. These were
+        // previously noncontextual checks that became contextual
+        // after Nu5 activation.
+        if (orchard_bundle.IsPresent()) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): transaction has Orchard actions"),
+                REJECT_INVALID, "bad-tx-has-orchard-actions");
+        }
```

Comment:
I'd argue that this is also an internal `zcashd` consistency check (since similarly to `nConsensusBranchId`, this can't occur for pre-v5 transactions from the network).

AFAICT we don't have an equivalent consistency check for Sapling, but I'm happy to have these here.